File: src/k6s/cli/session.py
"""Session management CLI commands."""

import asyncio
import json
from pathlib import Path

import typer
from rich.console import Console
from rich.panel import Panel
from rich.table import Table

from k6s.models.session import Session
from k6s.store.db import Database

app = typer.Typer()
console = Console()


@app.command(name="list")
def list_sessions(
    limit: int = typer.Option(20, "--limit", "-n", help="Maximum sessions to show"),
    all_states: bool = typer.Option(False, "--all", "-a", help="Show all states"),
) -> None:
    """List all sessions."""
    project_root = Path.cwd()
    khoregos_dir = project_root / ".khoregos"

    if not (khoregos_dir / "k6s.db").exists():
        console.print("[dim]No sessions found.[/dim]")
        return

    async def get_sessions() -> list[Session]:
        db = Database(khoregos_dir / "k6s.db")
        await db.connect()
        try:
            from k6s.engine.state import StateManager
            state_manager = StateManager(db, project_root)
            return await state_manager.list_sessions(limit=limit)
        finally:
            await db.close()

    sessions = asyncio.run(get_sessions())

    if not sessions:
        console.print("[dim]No sessions found.[/dim]")
        return

    table = Table(title="Sessions")
    table.add_column("ID", style="cyan")
    table.add_column("Objective")
    table.add_column("State")
    table.add_column("Started")
    table.add_column("Duration")
    table.add_column("Cost")
    table.add_column("Tokens")

    for session in sessions:
        duration = ""
        if session.duration_seconds:
            hours = int(session.duration_seconds // 3600)
            minutes = int((session.duration_seconds % 3600) // 60)
            duration = f"{hours}h {minutes}m" if hours else f"{minutes}m"

        state_style = {
            "completed": "green",
            "active": "yellow",
            "created": "yellow",
            "paused": "blue",
            "failed": "red",
        }.get(session.state.value, "white")

        tokens = ""
        if session.total_input_tokens or session.total_output_tokens:
            tokens = f"{session.total_input_tokens + session.total_output_tokens:,}"

        table.add_row(
            session.id[:8] + "...",
            session.objective[:35] + ("..." if len(session.objective) > 35 else ""),
            f"[{state_style}]{session.state.value}[/{state_style}]",
            session.started_at.strftime("%Y-%m-%d %H:%M"),
            duration or "-",
            f"${session.total_cost_usd:.2f}" if session.total_cost_usd else "-",
            tokens or "-",
        )

    console.print(table)


@app.command(name="latest")
def latest_session() -> None:
    """Show the most recent session (shortcut for `session show latest`)."""
    show_session("latest")


@app.command(name="show")
def show_session(
    session_id: str = typer.Argument(..., help="Session ID (or prefix). Use 'latest' for most recent."),
) -> None:
    """Show detailed session information."""
    project_root = Path.cwd()
    khoregos_dir = project_root / ".khoregos"

    if not (khoregos_dir / "k6s.db").exists():
        console.print("[yellow]No sessions found.[/yellow]")
        return

    async def get_session_details() -> dict | None:
        db = Database(khoregos_dir / "k6s.db")
        await db.connect()
        try:
            from k6s.engine.state import StateManager
            from k6s.engine.audit import AuditLogger

            state_manager = StateManager(db, project_root)

            # Find session: "latest" = most recent, otherwise match by ID or prefix
            sessions = await state_manager.list_sessions(limit=100)
            session = None
            if session_id.lower() == "latest" and sessions:
                session = sessions[0]
            else:
                for s in sessions:
                    if s.id.startswith(session_id) or s.id == session_id:
                        session = s
                        break

            if not session:
                return None

            # Get agents
            agents = await state_manager.list_agents(session.id)

            # Get audit event count
            audit_logger = AuditLogger(db, session.id)
            event_count = await audit_logger.get_event_count()

            # Get context entries
            context = await state_manager.load_all_context(session.id)

            return {
                "session": session,
                "agents": agents,
                "event_count": event_count,
                "context": context,
            }
        finally:
            await db.close()

    data = asyncio.run(get_session_details())

    if not data:
        console.print(f"[red]Session not found: {session_id}[/red]")
        raise typer.Exit(1)

    session = data["session"]
    agents = data["agents"]
    event_count = data["event_count"]
    context = data["context"]

    # Session info
    console.print(Panel(
        f"[bold]ID:[/bold] {session.id}\n"
        f"[bold]Objective:[/bold] {session.objective}\n"
        f"[bold]State:[/bold] {session.state.value}\n"
        f"[bold]Started:[/bold] {session.started_at.strftime('%Y-%m-%d %H:%M:%S')}\n"
        f"[bold]Ended:[/bold] {session.ended_at.strftime('%Y-%m-%d %H:%M:%S') if session.ended_at else '-'}\n"
        f"[bold]Parent:[/bold] {session.parent_session_id[:8] + '...' if session.parent_session_id else '-'}\n"
        f"[bold]Cost:[/bold] ${session.total_cost_usd:.4f}\n"
        f"[bold]Tokens:[/bold] {session.total_input_tokens:,} in / {session.total_output_tokens:,} out\n"
        f"[bold]Audit Events:[/bold] {event_count:,}",
        title="Session Details",
    ))

    # Agents
    if agents:
        console.print()
        console.print("[bold]Agents:[/bold]")
        for agent in agents:
            spec = f" ({agent.specialization})" if agent.specialization else ""
            console.print(f"  [cyan]{agent.name}[/cyan]{spec} - {agent.role.value}, {agent.state.value}")

    # Context summary
    if context:
        console.print()
        console.print("[bold]Saved Context:[/bold]")
        for entry in context[:10]:
            value_str = str(entry.value)
            if len(value_str) > 60:
                value_str = value_str[:60] + "..."
            console.print(f"  [dim]{entry.key}:[/dim] {value_str}")

    # Session summary
    if session.context_summary:
        console.print()
        console.print(Panel(session.context_summary, title="Session Summary"))


@app.command()
def context(
    session_id: str = typer.Argument(..., help="Session ID (or prefix). Use 'latest' for most recent."),
    key: str = typer.Option(None, "--key", "-k", help="Specific context key to show"),
    format: str = typer.Option("text", "--format", "-f", help="Output format: text, json"),
) -> None:
    """View saved context for a session."""
    project_root = Path.cwd()
    khoregos_dir = project_root / ".khoregos"

    if not (khoregos_dir / "k6s.db").exists():
        console.print("[yellow]No sessions found.[/yellow]")
        return

    async def get_context() -> list:
        db = Database(khoregos_dir / "k6s.db")
        await db.connect()
        try:
            from k6s.engine.state import StateManager
            state_manager = StateManager(db, project_root)

            # Find session: "latest" = most recent, otherwise match by ID or prefix
            sessions = await state_manager.list_sessions(limit=100)
            session = None
            if session_id.lower() == "latest" and sessions:
                session = sessions[0]
            else:
                for s in sessions:
                    if s.id.startswith(session_id) or s.id == session_id:
                        session = s
                        break

            if not session:
                return []

            if key:
                entry = await state_manager.load_context(session.id, key)
                return [entry] if entry else []
            else:
                return await state_manager.load_all_context(session.id)
        finally:
            await db.close()

    entries = asyncio.run(get_context())

    if not entries:
        if key:
            console.print(f"[yellow]Context key not found: {key}[/yellow]")
        else:
            console.print("[dim]No context saved for this session.[/dim]")
        return

    if format == "json":
        data = [{"key": e.key, "value": e.value, "updated_at": e.updated_at.isoformat()} for e in entries]
        print(json.dumps(data, indent=2))
    else:
        for entry in entries:
            console.print(f"[bold cyan]{entry.key}[/bold cyan]")
            console.print(f"[dim]Updated: {entry.updated_at.strftime('%Y-%m-%d %H:%M:%S')}[/dim]")
            console.print()

            if isinstance(entry.value, dict):
                console.print(json.dumps(entry.value, indent=2))
            else:
                console.print(str(entry.value))
            console.print()


@app.command()
def delete(
    session_id: str = typer.Argument(..., help="Session ID to delete"),
    force: bool = typer.Option(False, "--force", "-f", help="Skip confirmation"),
) -> None:
    """Delete a session and all its data."""
    project_root = Path.cwd()
    khoregos_dir = project_root / ".khoregos"

    if not (khoregos_dir / "k6s.db").exists():
        console.print("[yellow]No sessions found.[/yellow]")
        return

    async def get_session() -> Session | None:
        db = Database(khoregos_dir / "k6s.db")
        await db.connect()
        try:
            from k6s.engine.state import StateManager
            state_manager = StateManager(db, project_root)

            sessions = await state_manager.list_sessions(limit=100)
            for s in sessions:
                if s.id.startswith(session_id) or s.id == session_id:
                    return s
            return None
        finally:
            await db.close()

    session = asyncio.run(get_session())

    if not session:
        console.print(f"[red]Session not found: {session_id}[/red]")
        raise typer.Exit(1)

    if not force:
        console.print(f"[yellow]Delete session {session.id[:8]}...?[/yellow]")
        console.print(f"Objective: {session.objective}")
        confirm = typer.confirm("Are you sure?")
        if not confirm:
            console.print("[dim]Cancelled.[/dim]")
            raise typer.Exit(0)

    async def delete_session() -> None:
        db = Database(khoregos_dir / "k6s.db")
        await db.connect()
        try:
            # Delete related records first (foreign key constraints)
            await db.delete("audit_events", "session_id = ?", (session.id,))
            await db.delete("agents", "session_id = ?", (session.id,))
            await db.delete("context_store", "session_id = ?", (session.id,))
            await db.delete("file_locks", "session_id = ?", (session.id,))
            await db.delete("boundary_violations", "session_id = ?", (session.id,))
            await db.delete("gates", "session_id = ?", (session.id,))
            await db.delete("cost_records", "session_id = ?", (session.id,))
            await db.delete("sessions", "id = ?", (session.id,))
        finally:
            await db.close()

    asyncio.run(delete_session())
    console.print(f"[green]✓[/green] Session {session.id[:8]}... deleted")

File: src/cli/session.ts
/**
 * Session management CLI commands.
 */

import { existsSync } from "node:fs";
import path from "node:path";
import { Command } from "commander";
import chalk from "chalk";
import Table from "cli-table3";
import { Db } from "../store/db.js";
import { StateManager } from "../engine/state.js";
import { AuditLogger } from "../engine/audit.js";
import { type Session, sessionDurationSeconds } from "../models/session.js";

function withDb<T>(projectRoot: string, fn: (db: Db) => T): T {
  const db = new Db(path.join(projectRoot, ".khoregos", "k6s.db"));
  db.connect();
  try {
    return fn(db);
  } finally {
    db.close();
  }
}

function findSession(sm: StateManager, sessionId: string): Session | null {
  if (sessionId.toLowerCase() === "latest") {
    const sessions = sm.listSessions({ limit: 1 });
    return sessions[0] ?? null;
  }
  const sessions = sm.listSessions({ limit: 100 });
  for (const s of sessions) {
    if (s.id === sessionId || s.id.startsWith(sessionId)) return s;
  }
  return null;
}

export function registerSessionCommands(program: Command): void {
  const session = program
    .command("session")
    .description("Manage sessions");

  session
    .command("list")
    .description("List all sessions")
    .option("-n, --limit <number>", "Maximum sessions to show", "20")
    .action((opts: { limit: string }) => {
      const projectRoot = process.cwd();
      if (!existsSync(path.join(projectRoot, ".khoregos", "k6s.db"))) {
        console.log(chalk.dim("No sessions found."));
        return;
      }

      const sessions = withDb(projectRoot, (db) => {
        const sm = new StateManager(db, projectRoot);
        return sm.listSessions({ limit: parseInt(opts.limit, 10) });
      });

      if (!sessions.length) {
        console.log(chalk.dim("No sessions found."));
        return;
      }

      const table = new Table({
        head: ["ID", "Objective", "State", "Started", "Duration", "Cost", "Tokens"],
      });

      const stateColor: Record<string, (s: string) => string> = {
        completed: chalk.green,
        active: chalk.yellow,
        created: chalk.yellow,
        paused: chalk.blue,
        failed: chalk.red,
      };

      for (const s of sessions) {
        const dur = sessionDurationSeconds(s);
        let duration = "-";
        if (dur !== null) {
          const hours = Math.floor(dur / 3600);
          const minutes = Math.floor((dur % 3600) / 60);
          duration = hours ? `${hours}h ${minutes}m` : `${minutes}m`;
        }

        const colorFn = stateColor[s.state] ?? chalk.dim;
        const tokens =
          s.totalInputTokens || s.totalOutputTokens
            ? (s.totalInputTokens + s.totalOutputTokens).toLocaleString()
            : "-";

        table.push([
          s.id.slice(0, 8) + "...",
          s.objective.length > 35 ? s.objective.slice(0, 35) + "..." : s.objective,
          colorFn(s.state),
          new Date(s.startedAt).toISOString().slice(0, 16).replace("T", " "),
          duration,
          s.totalCostUsd ? `$${s.totalCostUsd.toFixed(2)}` : "-",
          tokens,
        ]);
      }

      console.log(table.toString());
    });

  session
    .command("latest")
    .description("Show the most recent session")
    .action(() => {
      showSessionDetails("latest");
    });

  session
    .command("show")
    .description("Show detailed session information")
    .argument("<session-id>", "Session ID (or prefix). Use 'latest' for most recent.")
    .action((sessionId: string) => {
      showSessionDetails(sessionId);
    });

  session
    .command("context")
    .description("View saved context for a session")
    .argument("<session-id>", "Session ID (or prefix). Use 'latest' for most recent.")
    .option("-k, --key <key>", "Specific context key to show")
    .option("-f, --format <format>", "Output format: text, json", "text")
    .action((sessionId: string, opts: { key?: string; format: string }) => {
      const projectRoot = process.cwd();
      if (!existsSync(path.join(projectRoot, ".khoregos", "k6s.db"))) {
        console.log(chalk.yellow("No sessions found."));
        return;
      }

      const entries = withDb(projectRoot, (db) => {
        const sm = new StateManager(db, projectRoot);
        const s = findSession(sm, sessionId);
        if (!s) return null;

        if (opts.key) {
          const entry = sm.loadContext(s.id, opts.key);
          return entry ? [entry] : [];
        }
        return sm.loadAllContext(s.id);
      });

      if (entries === null) {
        console.error(chalk.red(`Session not found: ${sessionId}`));
        process.exit(1);
      }

      if (!entries.length) {
        console.log(opts.key
          ? chalk.yellow(`Context key not found: ${opts.key}`)
          : chalk.dim("No context saved for this session."));
        return;
      }

      if (opts.format === "json") {
        const data = entries.map((e) => ({
          key: e.key,
          value: e.value,
          updated_at: e.updatedAt,
        }));
        console.log(JSON.stringify(data, null, 2));
      } else {
        for (const entry of entries) {
          console.log(chalk.cyan.bold(entry.key));
          console.log(chalk.dim(`Updated: ${new Date(entry.updatedAt).toISOString().slice(0, 19).replace("T", " ")}`));
          console.log();
          console.log(entry.value);
          console.log();
        }
      }
    });

  session
    .command("delete")
    .description("Delete a session and all its data")
    .argument("<session-id>", "Session ID to delete")
    .option("-f, --force", "Skip confirmation")
    .action((sessionId: string, opts: { force?: boolean }) => {
      const projectRoot = process.cwd();
      if (!existsSync(path.join(projectRoot, ".khoregos", "k6s.db"))) {
        console.log(chalk.yellow("No sessions found."));
        return;
      }

      const session = withDb(projectRoot, (db) => {
        const sm = new StateManager(db, projectRoot);
        return findSession(sm, sessionId);
      });

      if (!session) {
        console.error(chalk.red(`Session not found: ${sessionId}`));
        process.exit(1);
      }

      if (!opts.force) {
        // In a real CLI we'd prompt for confirmation.
        // For now, require --force.
        console.log(chalk.yellow(`Delete session ${session.id.slice(0, 8)}...?`));
        console.log(`Objective: ${session.objective}`);
        console.log("Use --force to confirm deletion.");
        return;
      }

      withDb(projectRoot, (db) => {
        const tables = [
          "audit_events",
          "agents",
          "context_store",
          "file_locks",
          "boundary_violations",
          "gates",
          "cost_records",
        ];
        for (const table of tables) {
          db.delete(table, "session_id = ?", [session.id]);
        }
        db.delete("sessions", "id = ?", [session.id]);
      });

      console.log(chalk.green("✓") + ` Session ${session.id.slice(0, 8)}... deleted`);
    });
}

function showSessionDetails(sessionId: string): void {
  const projectRoot = process.cwd();
  if (!existsSync(path.join(projectRoot, ".khoregos", "k6s.db"))) {
    console.log(chalk.yellow("No sessions found."));
    return;
  }

  const data = withDb(projectRoot, (db) => {
    const sm = new StateManager(db, projectRoot);
    const s = findSession(sm, sessionId);
    if (!s) return null;

    const agents = sm.listAgents(s.id);
    const al = new AuditLogger(db, s.id);
    const eventCount = al.getEventCount();
    const context = sm.loadAllContext(s.id);

    return { session: s, agents, eventCount, context };
  });

  if (!data) {
    console.error(chalk.red(`Session not found: ${sessionId}`));
    process.exit(1);
  }

  const { session, agents, eventCount, context } = data;

  console.log(chalk.bold("Session Details"));
  console.log(`  ${chalk.bold("ID:")} ${session.id}`);
  console.log(`  ${chalk.bold("Objective:")} ${session.objective}`);
  console.log(`  ${chalk.bold("State:")} ${session.state}`);
  console.log(`  ${chalk.bold("Started:")} ${new Date(session.startedAt).toISOString().slice(0, 19).replace("T", " ")}`);
  console.log(`  ${chalk.bold("Ended:")} ${session.endedAt ? new Date(session.endedAt).toISOString().slice(0, 19).replace("T", " ") : "-"}`);
  console.log(`  ${chalk.bold("Parent:")} ${session.parentSessionId ? session.parentSessionId.slice(0, 8) + "..." : "-"}`);
  console.log(`  ${chalk.bold("Cost:")} $${session.totalCostUsd.toFixed(4)}`);
  console.log(`  ${chalk.bold("Tokens:")} ${session.totalInputTokens.toLocaleString()} in / ${session.totalOutputTokens.toLocaleString()} out`);
  console.log(`  ${chalk.bold("Audit Events:")} ${eventCount.toLocaleString()}`);

  if (agents.length) {
    console.log();
    console.log(chalk.bold("Agents:"));
    for (const agent of agents) {
      const spec = agent.specialization ? ` (${agent.specialization})` : "";
      console.log(`  ${chalk.cyan(agent.name)}${spec} - ${agent.role}, ${agent.state}`);
    }
  }

  if (context.length) {
    console.log();
    console.log(chalk.bold("Saved Context:"));
    for (const entry of context.slice(0, 10)) {
      const val = entry.value.length > 60 ? entry.value.slice(0, 60) + "..." : entry.value;
      console.log(`  ${chalk.dim(entry.key + ":")} ${val}`);
    }
  }

  if (session.contextSummary) {
    console.log();
    console.log(chalk.bold("Session Summary:"));
    console.log(session.contextSummary);
  }
}
